File: Editorial/0107_Binary_Tree_Level_Order_Traversal_II/Go/2_Iteration_BFS_Traversal.go
func levelOrderBottom(root *TreeNode) [][]int {
    levels := make([][]int, 0)
    if root == nil {
        return levels
    }
    nextLevel := []*TreeNode{root}
    for len(nextLevel) > 0 {
        currLevel := nextLevel
        nextLevel = make([]*TreeNode, 0)
        levels = append(levels, make([]int, 0))
        for _, node := range currLevel {
            // append the current node value
            levels[len(levels)-1] = append(levels[len(levels)-1], node.Val)
            // process child nodes for the next level
            if node.Left != nil {
                nextLevel = append(nextLevel, node.Left)
            }
            if node.Right != nil {
                nextLevel = append(nextLevel, node.Right)
            }
        }
    }
    // reverse the array in-place
    for i := 0; i < len(levels)/2; i++ {
        levels[i], levels[len(levels)-1-i] = levels[len(levels)-1-i], levels[i]
    }
    return levels
}

File: Editorial/0107_Binary_Tree_Level_Order_Traversal_II/TypeScript/2_Iteration_BFS_Traversal.ts
function levelOrderBottom(root: TreeNode | null): number[][] {
    let levels: number[][] = [];
    if (root == null) return levels;
    let nextLevel: (TreeNode | null)[] = [root];
    while (nextLevel.length > 0) {
        let currLevel: (TreeNode | null)[] = [...nextLevel];
        nextLevel = [];
        levels.push([]);
        for (let node of currLevel) {
            if (node != null) {
                // append the current node value
                levels[levels.length - 1].push(node.val);
                // process child nodes for the next level
                if (node.left != null) nextLevel.push(node.left);
                if (node.right != null) nextLevel.push(node.right);
            }
        }
    }
    levels.reverse();
    return levels;
}
